ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,507 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+     */
+    if (QuotaType != PsPageFile &&
+        QuotaBlock != &PspDefaultQuotaBlock)
     {
-        InterlockedExchangeAdd((LONG*)&Process->QuotaUsage[QuotaType],
-                               -(LONG)Amount);
+        /* The return threshold can be non paged or paged */
+        ReturnThreshold = QuotaType ? PSP_NON_PAGED_POOL_QUOTA_THRESHOLD : PSP_PAGED_POOL_QUOTA_THRESHOLD;
+
+        /* Does the amount to return exceeds the threshold? */
+        if (Amount > ReturnThreshold)
+        {
+            /* It does, the new amount is the default */
+            AmountToReturn = ReturnThreshold;
+        }
+        else
+        {
+            /* Otherwise the amount is just the given one by the caller */
+            AmountToReturn = Amount;
+        }
+
+        /* Return the amounting we have back to Memory Manager */
+        InterlockedExchangeAddSizeT(QuotaBlock->QuotaEntry[QuotaType].Return, AmountToReturn);
+        MmReturnPoolQuota(QuotaType, QuotaBlock->QuotaEntry[QuotaType].Return);
+
+        /* And try to trim the limit */
+        InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Limit,
+                                        QuotaBlock->QuotaEntry[QuotaType].Limit - AmountToReturn,
+                                        QuotaBlock->QuotaEntry[QuotaType].Limit);
     }
 
+    /* Update the usage member of the block */
+    UpdatedQuotaUsage = QuotaBlock->QuotaEntry[QuotaType].Usage - Amount;
+    InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Usage,
+                                    UpdatedQuotaUsage,
+                                    QuotaBlock->QuotaEntry[QuotaType].Usage);
+
+    /* Are we being given a process? */
+    if (Process)
+    {
+        /* We're being given, check that's not a system one  */
+        ASSERT(Process != PsInitialSystemProcess);
+
+        /* Decrease the process' quota usage */
+        InterlockedExchangeAddSizeT(&Process->QuotaUsage[QuotaType], -(LONG)Amount);
+    }
+
+    /* We're done, release the lock */
     KeReleaseSpinLock(&PspQuotaLock, OldIrql);
 }
 
 /* FUNCTIONS ***************************************************************/
 
+/**
+ * @brief
+ * Initializes the quota system during boot
+ * phase of the system, which sets up the
+ * default quota block that is used across
+ * several processes.
+ *
+ * @return
+ * Nothing.
+ */
 CODE_SEG("INIT")
 VOID
 NTAPI
 PsInitializeQuotaSystem(VOID)
 {
+    /* Zero out the block */
     RtlZeroMemory(&PspDefaultQuotaBlock, sizeof(PspDefaultQuotaBlock));
+
+    /* Assign the default quota limits */
     PspDefaultQuotaBlock.QuotaEntry[PsNonPagedPool].Limit = (SIZE_T)-1;
     PspDefaultQuotaBlock.QuotaEntry[PsPagedPool].Limit = (SIZE_T)-1;
     PspDefaultQuotaBlock.QuotaEntry[PsPageFile].Limit = (SIZE_T)-1;
+
+    /*
+     * Set up the count references as the
+     * default block will gonna be used.
+     */
+    PspDefaultQuotaBlock.ReferenceCount = 1;
```

Comment:
Is the value "zero" meant to say the block has to go away?

Reply:
If `ReferenceCount == 0` then yes, the block is no longer used by anybody and it has to go away. Here the default quota block gets initialized upon boot phase of the system, gets assigned to the initial process yet we're not starting the counters which feels wrong.